### ISSUE TITLE
Call to undefined method ...\HasMany::byFeatureSlug() 

### ISSUE THREAD

**Issue by armandohg** [rinvex/laravel-subscriptions#62]
I'm getting this error when calling getFeatureRemainings()

Am I missing something?

**Comment by brunohccz** [comment]
I too.. Call to undefined method Illuminate\Database\Eloquent\Relations\HasMany::byFeatureSlug()

**Comment by karol-pietrzak** [comment]
I've managed to fix it i think. For now it runs ok for me. Scope `byFeatureSlug` seems to be not implemented but it is tried to be used.

Create two classes `PlanSubscriptionUsage` and `PlanSubscription` which extend those from package

**app/Models/PlanSubscriptionUsage.php** 
implementation of the missing scope
```
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Rinvex\Subscriptions\Models\PlanFeature;

class PlanSubscriptionUsage extends \Rinvex\Subscriptions\Models\PlanSubscriptionUsage
{

    /**
     * Scope subscription usage by feature slug.
     *
     * @param \Illuminate\Database\Eloquent\Builder $builder
     * @param string                                $featureSlug
     *
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public function scopeByFeatureSlug(Builder $builder, $featureSlug)
    {
        $feature = PlanFeature::where('slug', $featureSlug)->first();
        return $builder->where('feature_id', $feature->getKey() ?? null);
    }

}

```

**app/Models/PlanSubscription.php**
some extra tweaks to handle case if feature was never used
```
<?php


namespace App\Models;

class PlanSubscription extends \Rinvex\Subscriptions\Models\PlanSubscription
{
    /**
     * Determine if the feature can be used.
     *
     * @param string $featureSlug
     *
     * @return bool
     */
    public function canUseFeature(string $featureSlug): bool
    {
        $featureValue = $this->getFeatureValue($featureSlug);
        $usage = $this->usage()->byFeatureSlug($featureSlug)->first();

        if ($featureValue === 'true') {
            return true;
        }

        if(!$usage){
            $usage = new PlanSubscriptionUsage();
        }

        // If the feature value is zero, let's return false since
        // there's no uses available. (useful to disable countable features)
        if ($usage->expired() || is_null($featureValue) || $featureValue === '0' || $featureValue === 'false') {
            return false;
        }

        // Check for available uses
        return $this->getFeatureRemainings($featureSlug) > 0;
    }

    /**
     * Get how many times the feature has been used.
     *
     * @param string $featureSlug
     *
     * @return int
     */
    public function getFeatureUsage(string $featureSlug): int
    {
        $usage = $this->usage()->byFeatureSlug($featureSlug)->first();

        if(!$usage){
           return 0;
        }

        return ! $usage->expired() ? $usage->used : 0;
    }

}
```

Also very important to tell subscription package to use your own classes 

**config/rinvex.subscriptions.php**
```
<?php

declare(strict_types=1);

return [

    // Subscriptions Database Tables
    'tables' => [

        'plans' => 'plans',
        'plan_features' => 'plan_features',
        'plan_subscriptions' => 'plan_subscriptions',
        'plan_subscription_usage' => 'plan_subscription_usage',

    ],

    // Subscriptions Models
    'models' => [

        'plan' => \Rinvex\Subscriptions\Models\Plan::class,
        'plan_feature' => \Rinvex\Subscriptions\Models\PlanFeature::class,
        'plan_subscription' => App\Models\PlanSubscription::class,
        'plan_subscription_usage' => App\Models\PlanSubscriptionUsage::class,

    ],

];

```

I would be glad if creators patched that. Great simple package with some issues :)

**Comment by AlexHupe** [comment]
If $featureSlug cannot find in the database, an error occures: 

Call to a member function getKey() on null

**Comment by cheryllium** [comment]
Is the PR fixing this issue going to be merged? Still encountering this issue with v4.

**Comment by ziachand** [comment]
also getting this issue

**Comment by duongtdvn** [comment]
> If $featureSlug cannot find in the database, an error occures:
> 
> Call to a member function getKey() on null

this line should be replaced:
```php
$usage = $this->usage()->byFeatureSlug($featureSlug)->first();
```
with:
```php
$usage = $this->usage()->byFeatureSlug($featureSlug)->firstOrFail();
```

**Comment by andriasmelianus** [comment]
Hi @karol-pietrzak, thanks for the fix!
I found this part in `PlanSubscriptionUsage` is better like this:
````php
public function scopeByFeatureSlug(Builder $builder, $featureSlug)
{
    $feature = PlanFeature::where('slug', $featureSlug)->first();
    return $builder->where('feature_id', $feature ? $feature->getKey() : null);
}
````
Just in case we need to check a feature which does not exist under a plan, and we need a `FALSE` value in return.
And avoid error: `Call to a member function getKey() on null...`

**Comment by dranes** [comment]
It is valid to use something more SQLish?

```
public function scopeByFeatureSlug(Builder $builder, $featureSlug)
{

  return $builder->join(
              app('rinvex.subscriptions.plan_feature')->getTable(), 
              'feature_id' , 
              '=', 
              app('rinvex.subscriptions.plan_feature')->getTable() . "." . app('rinvex.subscriptions.plan_feature')->getKeyName()
          )->where('slug', $featureSlug);
}
```

**Comment by Omranic** [comment]
Similar to #43 #48 #62 #65 #136 #137 and was fixed previously on 0714e73, the fix already exists on the `develop` branch.

I'll tag a new version to merge the fix from `develop` to `master` 👍 

Sorry for the delay, and thank you for your patience!

@armandohg
@brunohccz
@karol-pietrzak
@AlexHupe
@cheryllium
@ziarv
@duongtdvn
@andriasmelianus
@dranes

Check here for details 
https://github.com/rinvex/laravel-subscriptions/blob/c11e6bb85fcd5dcf706cb7544e0d375654868bbb/src/Models/PlanSubscriptionUsage.php#L132-L137
